### PR TITLE
Migrate to Kotlin

### LINKED ISSUE #22: Rewrite in Kotlin
Kotlin-ify the library

### PR DESCRIPTION
**Pull request by hzsweers**
This migrates ViewPump to Kotlin, while leaving tests and the sample in Java to catch any potential API regressions

Resolves #22

There are some potentially breaking changes we could look at making, but separate decision once ViewPump 2.0 decisions are made.

The `-` prefixes on internal classes are to hide them from java, as they're otherwise visible as public classes

### CODE REVIEW

File: viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    fun invokeMethod(target: Any, method: Method?, vararg args: Any) {
+      try {
+        if (method == null) return
+        method.invoke(target, *args)
+      } catch (e: IllegalAccessException) {
+        Log.d(TAG, "Can't access method using reflection", e)
+      } catch (e: InvocationTargetException) {
+        Log.d(TAG, "Can't invoke method using reflection", e)
+      }
+
```
Comment (drakeet) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt]: Nit: This blank line is not needed.

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPump.kt [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+     * are created before reaching the LayoutInflater onViewCreated.
+     *
+     * @param enabled True if custom view inflated is allowed; otherwise, false.
+     */
+    fun setCustomViewInflationEnabled(enabled: Boolean) = apply {
+      this.customViewCreation = enabled
+    }
+
+    fun setReflectiveFallbackViewCreator(
+        reflectiveFallbackViewCreator: FallbackViewCreator) = apply {
```
Comment (drakeet) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ViewPump.kt]: Nit: No need to wrap here.

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+    return mutableView
+  }
+
+  private fun superOnCreateView(parent: View?, name: String, attrs: AttributeSet?): View? {
+    return try {
+      super.onCreateView(parent, name, attrs)
+    } catch (e: ClassNotFoundException) {
+      null
+    }
+
```
Comment (drakeet) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Nit: This blank line is not needed.

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+  }
+
+  private fun superOnCreateView(name: String, attrs: AttributeSet?): View? {
+    return try {
+      super.onCreateView(name, attrs)
+    } catch (e: ClassNotFoundException) {
+      null
+    }
+
```
Comment (drakeet) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Nit: This blank line is not needed.

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+      var view: View? = null
+      for (prefix in CLASS_PREFIX_LIST) {
+        try {
+          view = inflater.createView(name, prefix, attrs)
+          if (view != null) {
+            break
+          }
+        } catch (ignored: ClassNotFoundException) {
+        }
+
```
Comment (drakeet) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Nit: This blank line is not needed.

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+  companion object {
+
+    private val CLASS_PREFIX_LIST = setOf("android.widget.", "android.webkit.")
+    private val CONSTRUCTOR_ARGS_FIELD: Field by lazy {
+      requireNotNull(LayoutInflater::class.java.getDeclaredField("CONSTRUCTOR_ARGS_FIELD")) {
+        "No constructor arguments field found in LayoutInflater!"
+      }.apply { isAccessible = true }
+    }
+  }
+
```
Comment (drakeet) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Nit: This blank line is not needed.

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+  /**
+   * Factory 1 is the first port of call for LayoutInflation
+   */
+  private class WrapperFactory(factory: LayoutInflater.Factory) : LayoutInflater.Factory {
+
+    private val mViewCreator: FallbackViewCreator
+
+    init {
+      mViewCreator = WrapperFactoryViewCreator(factory)
```
Comment (jbarr21) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: inline assignment w/ definition?

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpContextWrapper.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * Remember if you are defining default in the
+ * [ViewPump] make sure this is initialised before
+ * the activity is created.
+ *
+ * @param base ContextBase to Wrap
+ */
+class ViewPumpContextWrapper private constructor(base: Context) : ContextWrapper(base) {
+
+  private var mInflater: ViewPumpLayoutInflater? = null
```
Comment (jbarr21) [inline viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpContextWrapper.kt]: lazy init and non-null?

**Comment by chrisjenx** [comment]
@hzsweers does this force java only clients to have to include the kotlin dep?

File: viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt (hunk 1 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt]: Trailing newline

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt (hunk 3 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Trailing newline

File: viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt (hunk 1 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt]: Trailing newline

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPump.kt [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+     *  If you disable this you will need to override your [android.app.Activity.onCreateView]
+     * as this is set as the [android.view.LayoutInflater] private factory.
+     * <br></br>
+     * ** Use the following code in the Activity if you disable FactoryInjection:**
+     * <pre>`
+     * @Override
+     * public View onCreateView(View parent, String name, Context context, AttributeSet attrs) {
+     * return ViewPumpContextWrapper.onActivityCreateView(this, parent, super.onCreateView(parent, name, context, attrs), name, context, attrs);
+     * }
+    `</pre> *
```
[the remark this answers is not in the document]
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPump.kt]: ef91fb2

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt (hunk 6 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Part of 8e74793

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt (hunk 5 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Part of 8e74793

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt (hunk 3 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Part of 8e74793

File: viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt (hunk 1 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ReflectionUtils.kt]: Part of 8e74793

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt (hunk 4 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: 0ead4dd

File: viewpump/src/main/java/io/github/inflationx/viewpump/-ViewPumpActivityFactory.kt [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+internal interface ViewPumpActivityFactory {
+
+  /**
+   * Used to Wrap the Activity onCreateView method.
+   *
+   * You implement this method like so in you base activity.
+   *
+   * ```
+   * public View onCreateView(View parent, String name, Context context, AttributeSet attrs) {
```
[the remark this answers is not in the document]
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/-ViewPumpActivityFactory.kt]: Listing both e78f9a0

File: viewpump/src/main/java/io/github/inflationx/viewpump/-ReflectiveFallbackViewCreator.kt [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    } catch (e: InstantiationException) {
+      e.printStackTrace()
+    } catch (e: InvocationTargetException) {
+      e.printStackTrace()
+    }
+
+    return null
+  }
+
+  companion object {
```
[the remark this answers is not in the document]
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/-ReflectiveFallbackViewCreator.kt]: 7094638

File: viewpump/src/main/java/io/github/inflationx/viewpump/-ReflectiveFallbackViewCreator.kt [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        constructor = clazz.getConstructor(*CONSTRUCTOR_SIGNATURE_2)
+        constructorArgs = arrayOf(context, attrs)
+      } catch (e: NoSuchMethodException) {
+        constructor = clazz.getConstructor(*CONSTRUCTOR_SIGNATURE_1)
+        constructorArgs = arrayOf(context)
+      }
+
+      constructor.isAccessible = true
+      return constructor.newInstance(*constructorArgs)
+    } catch (e: ClassNotFoundException) {
```
[the remark this answers is not in the document]
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/-ReflectiveFallbackViewCreator.kt]: Kotlin doesn't have multicatch 86c590a and e704b70

File: viewpump/src/main/java/io/github/inflationx/viewpump/-InterceptorChain.kt [hunk 13]
```diff
@@ -0,0 +1,25 @@
+@file:JvmName("InterceptorChain")
```
[the remark this answers is not in the document]
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/-InterceptorChain.kt]: Derp good catch 7829291

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+import android.view.View
+import android.view.ViewGroup
+
+import org.xmlpull.v1.XmlPullParser
+
+import java.lang.reflect.Field
+
+internal class ViewPumpLayoutInflater : LayoutInflater, ViewPumpActivityFactory {
+
+  private var nameAndAttrsViewCreator: FallbackViewCreator? = null
```
[the remark this answers is not in the document]
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: Can actually just make them `val` ec08f7d

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt (hunk 7 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpLayoutInflater.kt]: f330335

File: viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpContextWrapper.kt (hunk 8 above)
Comment (hzsweers) [reply viewpump/src/main/java/io/github/inflationx/viewpump/ViewPumpContextWrapper.kt]: 475dde5

**Comment by hzsweers** [comment]
> does this force java only clients to have to include the kotlin dep?

yeah it does

**Comment by hzsweers** [comment]
@drakeet thanks for the extra review!

**Pull request merged**